### ISSUE TITLE
Replace boost::filesystem with std::filesystem?

### ISSUE THREAD

**Issue by stuarteberg** [constantinpape/z5#123]
This code depends on `boost-cpp` at runtime due to its use of `boost::filesystem`.  But C++17 added `std::filesystem`, which is largely compatible with the boost implementation.  

Now that conda-forge supports C++17 on Linux and Mac, do you have any plans to replace `boost::filesystem` with `std::filesystem`?  (I'm not 100% sure about Windows support, so maybe you can create a compiler switch that allows users to compile with boost if necessary.)

As far as I can see, `z5` only uses the following `filesystem` functions, all of which are now available in `std::filesystem`:

```
$ grep -roh 'fs::\w*' src include | sort | uniq
fs::create_directories
fs::create_directory
fs::directory_iterator
fs::exists
fs::fstream
fs::ifstream
fs::is_regular_file
fs::ofstream
fs::path
fs::recursive_directory_iterator
fs::remove
fs::remove_all
```

Obviously, one advantage to this is that it might (hopefully) solve https://github.com/conda-forge/z5py-feedstock/issues/15

**Comment by jakirkham** [comment]
Are there other uses of Boost in the codebase or is that it?

**Comment by constantinpape** [comment]
>  do you have any plans to replace `boost::filesystem` with `std::filesystem`?

Yes, I have thought about this too. And it would make conda builds much easier if we could get rid of the boost dependency.
I know of several users who build this themselves on windows and probably don't have C++17 support (@weilewei, syglass (@JuanPotato / @mmoreha)).
So we can't completely get rid of boost completely yet.

> so maybe you can create a compiler switch that allows users to compile with boost if necessary

Yes, something like this should do the trick:
In CMakeLists
```cmake
if(not CPP17)
  find_package(BOOST Required)
  add_definitions(-DWITH_BOOST_FS)
endif()
```
and then in the code.
```cpp
#ifdef WITH_BOOST_FS
#include "boost/filesystem.hpp"
namespace fs = boost::filesystem;
#else
#include <filesystem>
namespace fs = std::filesystem;
#endif
```
Any other ideas how this should be done?

@jakirkham 
> Are there other uses of Boost in the codebase or is that it?

No, that's it.
While I have your attention, could I ask you to take a very quick look at my azure problems 
in the conda-forge build: https://github.com/conda-forge/z5py-feedstock/pull/13#issuecomment-518317090. Thanks!

**Comment by jakirkham** [comment]
Seems reasonable.

Cool, agree with Stuart, making Boost a dependency only for legacy compilers would be quite nice.

Windows probably will need VS 2017 to avoid the Boost dependency entirely. Though falling back to Boost should be sufficient here.

Sure, [done]( https://github.com/conda-forge/z5py-feedstock/pull/13#issuecomment-518603234 ).

**Comment by constantinpape** [comment]
Ok, I will look into making the boost dependency optional for legacy compilers.
If anyone has a better idea than this:

> Yes, something like this should do the trick:
> In CMakeLists
> 
> ```cmake
> if(not CPP17)
>   find_package(BOOST Required)
>   add_definitions(-DWITH_BOOST_FS)
> endif()
> ```
> 
> and then in the code.
> 
> ```c++
> #ifdef WITH_BOOST_FS
> #include "boost/filesystem.hpp"
> namespace fs = boost::filesystem;
> #else
> #include <filesystem>
> namespace fs = std::filesystem;
> #endif
> ```

let me know.

**Comment by stuarteberg** [comment]
>Are there other uses of Boost in the codebase or is that it?

Well, there are a couple of header-only dependencies:

```
$ grep -roh '<boost/.*>' include src | sort | uniq
<boost/any.hpp>
<boost/filesystem.hpp>
<boost/filesystem/fstream.hpp>
<boost/iterator/counting_iterator.hpp>
<boost/iterator/transform_iterator.hpp>
```

...so you'll still need `boost-cpp` as a compile-time dependency.  But not as a run-time dependency, which is what I care about.

Regarding cmake:

> Yes, something like this should do the trick:
> In CMakeLists
> 
> ```cmake
> if(not CPP17)
>   find_package(BOOST Required)
>   add_definitions(-DWITH_BOOST_FS)
> endif()
> ```

I guess I'd prefer to give the user the ability to use boost if they really want to, even with a newer compiler.  Perhaps you could define a cache variable of the same name, so the user could write this if desired:

`cmake -DWITH_BOOST_FS=1 ..`

In `CMakeLists.txt`, you would need to ensure that `CPP17` would merely determine the cache variable's default value.

**Comment by constantinpape** [comment]
> Well, there are a couple of header-only dependencies:

Right, completely forgot about these.
But indeed, we will not need the run-time dependency for the header only parts and also won't need to pin the version then.

> I guess I'd prefer to give the user the ability to use boost if they really want to, even with a newer compiler. Perhaps you could define a cache variable of the same name, so the user could write this if desired:
> 
> `cmake -DWITH_BOOST_FS=1 ..`

That's a good point.

**Comment by constantinpape** [comment]
Implemented in #124.
